File: teaching/anthropic/anthropic-protocol-book/code/python/ch09/batch_client.py
"""
Chapter 9: BatchClient — async bulk processing via the Message Batches API.

Protocol-level implementation that communicates directly with the Anthropic
HTTP API (no SDK dependency). Covers the full MessageBatch lifecycle:
create → poll status → retrieve results → interpret per-request outcomes.

Verified against official docs (docs.anthropic.com, May 2026):
  - 50% cost discount on all models
  - 24-hour completion SLA (batches expire after 24h)
  - Results available for 29 days
  - Max 100 000 requests or 256 MB per batch
  - Per-request result types: succeeded, errored, canceled, expired
  - Prompt caching supported (5-min ephemeral TTL, best-effort)
  - 1-hour cache TTL available via cache_control ttl parameter
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional

import httpx


# ---------------------------------------------------------------------------
# Domain types
# ---------------------------------------------------------------------------


class BatchProcessingStatus(str, Enum):
    """MessageBatch lifecycle states.

    Per the Anthropic API: a batch starts in ``in_progress``, then transitions
    to ``ended`` once every request has been processed.  Users may also
    explicitly cancel a batch (``canceled``), or the system will expire it
    if processing exceeds 24 hours (``expired``).
    """

    IN_PROGRESS = "in_progress"
    ENDED = "ended"
    CANCELED = "canceled"
    EXPIRED = "expired"


class ResultType(str, Enum):
    """Per-request outcome within a completed batch.

    ``succeeded``
        The request was processed and the model returned a message.  You
        are billed for these requests at the batch rate.
    ``errored``
        The request could not be processed (invalid params, internal server
        error, etc.).  You are **not** billed.
    ``canceled``
        The batch was canceled before this request reached the model.
        You are **not** billed.
    ``expired``
        The 24-hour SLA elapsed before this request could be processed.
        You are **not** billed.
    """

    SUCCEEDED = "succeeded"
    ERRORED = "errored"
    CANCELED = "canceled"
    EXPIRED = "expired"


@dataclass
class MessageRequest:
    """A single request within a batch.

    Attributes:
        custom_id:
            Client-defined identifier.  Must be unique within the batch.
            Used to correlate results back to source requests since the
            response order is not guaranteed.
        params:
            Standard Messages API parameters (``model``, ``max_tokens``,
            ``messages``, and optionally ``system``, ``tools``, ``temperature``,
            ``top_p``, ``stop_sequences``, ``metadata``, etc.).
    """

    custom_id: str
    params: Dict[str, Any]


@dataclass
class BatchInfo:
    """Summary returned by the ``/v1/messages/batches/{id}`` endpoint."""

    id: str
    processing_status: BatchProcessingStatus
    request_counts: Dict[str, int] = field(default_factory=dict)
    created_at: str = ""
    ended_at: str = ""
    expires_at: str = ""
    results_url: Optional[str] = None
    raw: Dict[str, Any] = field(default_factory=dict)


@dataclass
class BatchResult:
    """Single per-request result extracted from the JSONL results stream."""

    custom_id: str
    result_type: ResultType
    message: Optional[Dict[str, Any]] = None  # Present when succeeded
    error: Optional[Dict[str, Any]] = None     # Present when errored
    raw: Dict[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# HTTP client
# ---------------------------------------------------------------------------

DEFAULT_BASE_URL = "https://api.anthropic.com"
ANTHROPIC_VERSION = "2023-06-01"


class BatchClient:
    """Asynchronous bulk processing via the Anthropic Message Batches API.

    Communicates directly over HTTP — no SDK dependency.  Designed for
    protocol-level education: every method maps 1:1 to a documented endpoint.

    Usage::

        client = BatchClient(api_key="sk-ant-...")
        batch_id = client.create_batch([
            MessageRequest(custom_id="q1", params={...}),
            MessageRequest(custom_id="q2", params={...}),
        ])
        results = client.wait_for_completion(batch_id)

    Reference endpoints
    -------------------
    =======  =====================================================
    Method    Endpoint
    =======  =====================================================
    POST      /v1/messages/batches
    GET       /v1/messages/batches/{message_batch_id}
    GET       /v1/messages/batches/{message_batch_id}/results
    POST      /v1/messages/batches/{message_batch_id}/cancel
    GET       /v1/messages/batches
    =======  =====================================================
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = DEFAULT_BASE_URL,
    ) -> None:
        self._api_key = api_key or os.environ.get("ANTHROPIC_API_KEY", "")
        self._base_url = base_url.rstrip("/")
        self._client = httpx.Client(
            base_url=self._base_url,
            headers={
                "x-api-key": self._api_key,
                "anthropic-version": ANTHROPIC_VERSION,
                "content-type": "application/json",
            },
            timeout=httpx.Timeout(120.0),
        )

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def create_batch(self, requests: List[MessageRequest]) -> str:
        """Create a new Message Batch and return its ``batch_id``.

        The batch starts in ``in_progress``.  Requests are processed
        asynchronously and independently — one failure does not affect others.
        """
        body = {
            "requests": [
                {"custom_id": r.custom_id, "params": r.params} for r in requests
            ]
        }
        resp = self._client.post("/v1/messages/batches", json=body)
        resp.raise_for_status()
        data = resp.json()
        return data["id"]

    def get_batch_status(self, batch_id: str) -> BatchInfo:
        """Retrieve the current state of a Message Batch."""
        resp = self._client.get(f"/v1/messages/batches/{batch_id}")
        resp.raise_for_status()
        data = resp.json()
        return BatchInfo(
            id=data["id"],
            processing_status=BatchProcessingStatus(data["processing_status"]),
            request_counts=data.get("request_counts", {}),
            created_at=data.get("created_at", ""),
            ended_at=data.get("ended_at", ""),
            expires_at=data.get("expires_at", ""),
            results_url=data.get("results_url"),
            raw=data,
        )

    def get_batch_results(self, batch_id: str) -> List[BatchResult]:
        """Download and parse a completed batch's JSONL results.

        Each line of the response is a JSON object representing one request's
        outcome.  The ``custom_id`` field allows you to correlate results
        back to the original requests.
        """
        resp = self._client.get(
            f"/v1/messages/batches/{batch_id}/results",
        )
        resp.raise_for_status()
        return [self._parse_result_line(line) for line in resp.text.strip().splitlines() if line.strip()]

    def cancel_batch(self, batch_id: str) -> BatchInfo:
        """Cancel an in-progress batch.

        Already-processed requests retain their outcomes; pending requests
        are marked ``canceled`` and you are not billed for them.
        """
        resp = self._client.post(
            f"/v1/messages/batches/{batch_id}/cancel",
        )
        resp.raise_for_status()
        data = resp.json()
        return BatchInfo(
            id=data["id"],
            processing_status=BatchProcessingStatus(data["processing_status"]),
            request_counts=data.get("request_counts", {}),
            created_at=data.get("created_at", ""),
            ended_at=data.get("ended_at", ""),
            expires_at=data.get("expires_at", ""),
            results_url=data.get("results_url"),
            raw=data,
        )

    def list_batches(self, limit: int = 20, after_id: Optional[str] = None) -> List[BatchInfo]:
        """List all batches in the workspace, newest first.

        Args:
            limit: Max batches to return (default 20, max 100).
            after_id: Cursor — return batches after this ID.
        """
        params: Dict[str, Any] = {"limit": limit}
        if after_id:
            params["after_id"] = after_id
        resp = self._client.get("/v1/messages/batches", params=params)
        resp.raise_for_status()
        data = resp.json()
        return [
            BatchInfo(
                id=b["id"],
                processing_status=BatchProcessingStatus(b["processing_status"]),
                request_counts=b.get("request_counts", {}),
                created_at=b.get("created_at", ""),
                ended_at=b.get("ended_at", ""),
                expires_at=b.get("expires_at", ""),
                results_url=b.get("results_url"),
                raw=b,
            )
            for b in data.get("data", [])
        ]

    def wait_for_completion(
        self,
        batch_id: str,
        poll_interval: int = 60,
        max_wait: int = 86400,
    ) -> List[BatchResult]:
        """Block until the batch finishes, then return results.

        Args:
            batch_id: The batch to wait for.
            poll_interval: Seconds between status checks (default 60).
            max_wait: Maximum total wait in seconds (default 86 400 = 24h).

        Returns:
            Parsed per-request results.

        Raises:
            TimeoutError: If the batch does not finish within ``max_wait``.
            RuntimeError: If the batch expires or is canceled.
        """
        deadline = time.monotonic() + max_wait
        terminal = {
            BatchProcessingStatus.ENDED,
            BatchProcessingStatus.CANCELED,
            BatchProcessingStatus.EXPIRED,
        }

        while time.monotonic() < deadline:
            info = self.get_batch_status(batch_id)
            if info.processing_status in terminal:
                break
            time.sleep(poll_interval)
        else:
            raise TimeoutError(
                f"Batch {batch_id} did not complete within {max_wait}s"
            )

        if info.processing_status == BatchProcessingStatus.ENDED:
            return self.get_batch_results(batch_id)

        if info.processing_status == BatchProcessingStatus.EXPIRED:
            raise RuntimeError(
                f"Batch {batch_id} expired (24h SLA exceeded)"
            )

        if info.processing_status == BatchProcessingStatus.CANCELED:
            raise RuntimeError(
                f"Batch {batch_id} was canceled"
            )

        return []  # unreachable

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _parse_result_line(line: str) -> BatchResult:
        data = json.loads(line)
        custom_id = data.get("custom_id", "")
        result = data.get("result", {})

        result_type_str = result.get("type", "")

        try:
            result_type = ResultType(result_type_str)
        except ValueError:
            result_type = ResultType.ERRORED

        return BatchResult(
            custom_id=custom_id,
            result_type=result_type,
            message=result.get("message") if result_type == ResultType.SUCCEEDED else None,
            error=result.get("error") if result_type == ResultType.ERRORED else None,
            raw=data,
        )

    # ------------------------------------------------------------------
    # Batch + Cache combo helpers
    # ------------------------------------------------------------------

    @staticmethod
    def make_cacheable_request(
        custom_id: str,
        model: str,
        max_tokens: int,
        user_message: str,
        system_prompt: str = "",
        cache_ttl: Optional[int] = None,  # seconds; None → 5-min default
    ) -> MessageRequest:
        """Factory for a request with prompt-caching enabled.

        When used inside a batch, the batch processing engine attempts to
        re-use cached prefixes across requests that share identical content,
        on a best-effort basis.  Cache hit rates typically range from 30-98%
        depending on traffic patterns.

        The 1-hour TTL (``cache_ttl=3600``) is useful when batch requests
        are spread out over a longer window, though note that batch cache
        behaviour is still best-effort.

        Args:
            custom_id: Unique request identifier.
            model: Model ID string.
            max_tokens: Max output tokens.
            user_message: The user's question / prompt.
            system_prompt: Optional system-level instructions (cached).
            cache_ttl: If provided, sets ``ttl`` in ``cache_control`` for
                       a 1-hour cache. Omit for default 5-min ephemeral.
        """
        system: list[dict[str, Any]] = []
        if system_prompt:
            cache_ctl: dict[str, Any] = {"type": "ephemeral"}
            if cache_ttl is not None:
                cache_ctl["ttl"] = cache_ttl
            system.append({
                "type": "text",
                "text": system_prompt,
                "cache_control": cache_ctl,
            })

        params: dict[str, Any] = {
            "model": model,
            "max_tokens": max_tokens,
            "messages": [{"role": "user", "content": user_message}],
        }
        if system:
            params["system"] = system

        return MessageRequest(custom_id=custom_id, params=params)

File: teaching/anthropic/anthropic-protocol-book/code/node/ch09/BatchClient.ts
/**
 * Chapter 9: BatchClient (TypeScript) — async bulk processing via the
 * Anthropic Message Batches API.
 *
 * Protocol-level implementation communicating directly over HTTP.
 * Covers the full MessageBatch lifecycle:
 *   create → poll status → retrieve results → interpret per-request outcomes.
 *
 * Verified against official docs (docs.anthropic.com, May 2026):
 *   - 50% cost discount on all models
 *   - 24-hour completion SLA (batches expire after 24h)
 *   - Results available for 29 days
 *   - Max 100 000 requests or 256 MB per batch
 *   - Per-request result types: succeeded, errored, canceled, expired
 *   - Prompt caching supported (5-min ephemeral TTL, best-effort)
 *   - 1-hour cache TTL available via cache_control ttl parameter
 */

// ---------------------------------------------------------------------------
// Domain types
// ---------------------------------------------------------------------------

export type BatchProcessingStatus =
  | "in_progress"
  | "ended"
  | "canceled"
  | "expired";

export type ResultType = "succeeded" | "errored" | "canceled" | "expired";

export interface MessageRequest {
  /** Client-defined unique identifier within the batch. */
  custom_id: string;
  /** Standard Messages API parameters (model, max_tokens, messages, …). */
  params: Record<string, unknown>;
}

export interface BatchInfo {
  id: string;
  processing_status: BatchProcessingStatus;
  request_counts: Record<string, number>;
  created_at: string;
  ended_at: string;
  expires_at: string;
  results_url: string | null;
  raw: Record<string, unknown>;
}

export interface BatchResult {
  custom_id: string;
  result_type: ResultType;
  /** Present when result_type is "succeeded". */
  message: Record<string, unknown> | null;
  /** Present when result_type is "errored". */
  error: Record<string, unknown> | null;
  raw: Record<string, unknown>;
}

export interface MakeCacheableRequestOptions {
  custom_id: string;
  model: string;
  max_tokens: number;
  user_message: string;
  system_prompt?: string;
  /** Seconds. If set, uses 1-hour TTL. Omit for default 5-min ephemeral. */
  cache_ttl?: number;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_BASE_URL = "https://api.anthropic.com";
const ANTHROPIC_VERSION = "2023-06-01";

// ---------------------------------------------------------------------------
// BatchClient
// ---------------------------------------------------------------------------

export class BatchClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(apiKey?: string, baseUrl: string = DEFAULT_BASE_URL) {
    this.apiKey = apiKey ?? process.env["ANTHROPIC_API_KEY"] ?? "";
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  // ------------------------------------------------------------------
  // Public API
  // ------------------------------------------------------------------

  /**
   * Create a new Message Batch and return its `batch_id`.
   *
   * The batch starts in `in_progress`.  Requests are processed
   * asynchronously and independently — one failure does not affect others.
   */
  async createBatch(requests: MessageRequest[]): Promise<string> {
    const body = {
      requests: requests.map((r) => ({
        custom_id: r.custom_id,
        params: r.params,
      })),
    };

    const resp = await fetch(`${this.baseUrl}/v1/messages/batches`, {
      method: "POST",
      headers: this._headers(),
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      throw new Error(
        `Create batch failed (${resp.status}): ${await resp.text()}`,
      );
    }

    const data = (await resp.json()) as { id: string };
    return data.id;
  }

  /**
   * Retrieve the current state of a Message Batch.
   */
  async getBatchStatus(batchId: string): Promise<BatchInfo> {
    const resp = await fetch(
      `${this.baseUrl}/v1/messages/batches/${batchId}`,
      { headers: this._headers() },
    );

    if (!resp.ok) {
      throw new Error(
        `Get batch status failed (${resp.status}): ${await resp.text()}`,
      );
    }

    const data = (await resp.json()) as Record<string, unknown>;
    return this._toBatchInfo(data);
  }

  /**
   * Download and parse a completed batch's JSONL results.
   *
   * Each line of the response is a JSON object representing one request's
   * outcome.  The `custom_id` field allows correlation back to original requests.
   */
  async getBatchResults(batchId: string): Promise<BatchResult[]> {
    const resp = await fetch(
      `${this.baseUrl}/v1/messages/batches/${batchId}/results`,
      { headers: this._headers() },
    );

    if (!resp.ok) {
      throw new Error(
        `Get batch results failed (${resp.status}): ${await resp.text()}`,
      );
    }

    const text = await resp.text();
    return text
      .trim()
      .split("\n")
      .filter((line) => line.trim().length > 0)
      .map((line) => this._parseResultLine(line));
  }

  /**
   * Cancel an in-progress batch.
   *
   * Already-processed requests retain their outcomes; pending requests
   * are marked `canceled` and you are not billed for them.
   */
  async cancelBatch(batchId: string): Promise<BatchInfo> {
    const resp = await fetch(
      `${this.baseUrl}/v1/messages/batches/${batchId}/cancel`,
      {
        method: "POST",
        headers: this._headers(),
      },
    );

    if (!resp.ok) {
      throw new Error(
        `Cancel batch failed (${resp.status}): ${await resp.text()}`,
      );
    }

    const data = (await resp.json()) as Record<string, unknown>;
    return this._toBatchInfo(data);
  }

  /**
   * List all batches in the workspace, newest first.
   */
  async listBatches(limit: number = 20, afterId?: string): Promise<BatchInfo[]> {
    const params = new URLSearchParams({ limit: String(limit) });
    if (afterId) params.set("after_id", afterId);

    const resp = await fetch(
      `${this.baseUrl}/v1/messages/batches?${params.toString()}`,
      { headers: this._headers() },
    );

    if (!resp.ok) {
      throw new Error(
        `List batches failed (${resp.status}): ${await resp.text()}`,
      );
    }

    const data = (await resp.json()) as { data?: Record<string, unknown>[] };
    return (data.data ?? []).map((b) => this._toBatchInfo(b));
  }

  /**
   * Block until the batch finishes, then return results.
   *
   * @param pollInterval - Seconds between status checks (default 60).
   * @param maxWait       - Maximum total wait in seconds (default 86 400 = 24h).
   * @throws If the batch expires, is canceled, or times out.
   */
  async waitForCompletion(
    batchId: string,
    pollInterval: number = 60,
    maxWait: number = 86400,
  ): Promise<BatchResult[]> {
    const deadline = Date.now() + maxWait * 1000;
    const terminal: Set<BatchProcessingStatus> = new Set([
      "ended",
      "canceled",
      "expired",
    ]);

    let info: BatchInfo | null = null;

    while (Date.now() < deadline) {
      info = await this.getBatchStatus(batchId);
      if (terminal.has(info.processing_status)) break;
      await this._sleep(pollInterval * 1000);
    }

    if (!info || !terminal.has(info.processing_status)) {
      throw new Error(
        `Batch ${batchId} did not complete within ${maxWait}s`,
      );
    }

    if (info.processing_status === "ended") {
      return this.getBatchResults(batchId);
    }

    if (info.processing_status === "expired") {
      throw new Error(`Batch ${batchId} expired (24h SLA exceeded)`);
    }

    throw new Error(`Batch ${batchId} was canceled`);
  }

  // ------------------------------------------------------------------
  // Batch + Cache combo helpers
  // ------------------------------------------------------------------

  /**
   * Factory for a request with prompt-caching enabled.
   *
   * When used inside a batch, the batch processing engine attempts to
   * re-use cached prefixes across requests that share identical content,
   * on a best-effort basis.  Cache hit rates typically range from 30-98%
   * depending on traffic patterns.
   *
   * The 1-hour TTL (``cache_ttl=3600``) is useful when batch requests
   * are spread out over a longer window, though batch cache behaviour
   * is still best-effort.
   */
  static makeCacheableRequest(opts: MakeCacheableRequestOptions): MessageRequest {
    const params: Record<string, unknown> = {
      model: opts.model,
      max_tokens: opts.max_tokens,
      messages: [{ role: "user", content: opts.user_message }],
    };

    if (opts.system_prompt) {
      const cacheControl: Record<string, unknown> = { type: "ephemeral" };
      if (opts.cache_ttl !== undefined) {
        cacheControl["ttl"] = opts.cache_ttl;
      }
      params["system"] = [
        {
          type: "text",
          text: opts.system_prompt,
          cache_control: cacheControl,
        },
      ];
    }

    return { custom_id: opts.custom_id, params };
  }

  // ------------------------------------------------------------------
  // Internal helpers
  // ------------------------------------------------------------------

  private _headers(): Record<string, string> {
    return {
      "x-api-key": this.apiKey,
      "anthropic-version": ANTHROPIC_VERSION,
      "content-type": "application/json",
    };
  }

  private _toBatchInfo(data: Record<string, unknown>): BatchInfo {
    return {
      id: String(data["id"] ?? ""),
      processing_status: String(
        data["processing_status"] ?? "in_progress",
      ) as BatchProcessingStatus,
      request_counts: (data["request_counts"] as Record<string, number>) ?? {},
      created_at: String(data["created_at"] ?? ""),
      ended_at: String(data["ended_at"] ?? ""),
      expires_at: String(data["expires_at"] ?? ""),
      results_url: data["results_url"] ? String(data["results_url"]) : null,
      raw: data,
    };
  }

  private _parseResultLine(line: string): BatchResult {
    const data = JSON.parse(line) as Record<string, unknown>;
    const customId = String(data["custom_id"] ?? "");
    const result = (data["result"] ?? {}) as Record<string, unknown>;
    const resultType = String(result["type"] ?? "errored") as ResultType;

    return {
      custom_id: customId,
      result_type: resultType,
      message:
        resultType === "succeeded"
          ? (result["message"] as Record<string, unknown> | null)
          : null,
      error:
        resultType === "errored"
          ? (result["error"] as Record<string, unknown> | null)
          : null,
      raw: data,
    };
  }

  private _sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
